api/v1beta1/rabbitmqcluster_types.go

```diff
@@ -64,6 +64,10 @@ type RabbitmqClusterSpec struct {
 	Rabbitmq    RabbitmqClusterConfigurationSpec `json:"rabbitmq,omitempty"`
 	TLS         TLSSpec                          `json:"tls,omitempty"`
 	Override    RabbitmqClusterOverrideSpec      `json:"override,omitempty"`
+	// If set to true, the cluster will run `rabbitmq-upgrade post_upgrade` whenever the cluster is updated.
+	// Has no effect if the cluster only consists of one node.
+	// For more information, see https://www.rabbitmq.com/rabbitmq-upgrade.8.html#post_upgrade
+	RunPostUpgradeSteps bool `json:"runPostUpgradeSteps,omitempty"`
```

Comment:
This operation is very disruptive (IIRC, no messages can be published/consumed) and the user should have the ability to decide when/whether to run this command. Note that there are few scenarios where RabbitMQ is upgraded and you don't want to run these steps, however the intention of this PR is provide the ability to run the command after a rolling restart, which does not necessarily happen in the context of an upgrade e.g. I change `rabbitmq.conf` and the cluster roll-restarts.

Reply:
In my mind, there's two reasons.

Practically, post_upgrade can be a resource-intensive operation, especially if there are many nodes or queues. You may not wish for this resource consumption after every upgrade, though I could also see the argument for this resource consumption being less than that caused by the upgrade itself, and having unbalanced queue leaders!

More widely, the operator is positioned as an unopinionated method of reliably deploying RabbitMQ in Kubernetes. I agree, I think this *should* be run after every cluster upgrade, and when I originally started this I didn't have a switch at all, let alone default it to false. But I think this provides a more typical out-of-the-box RabbitMQ cluster experience when set to defaults this way than if it was always run.

That said, if rebalancing the queues every time is shown to be minimally disruptive compared to not rebalancing the queues, then I'd be in favour of making this a 'Skip' instead of a 'Run' flag, so that post_upgrade is run by default but configurably not run.

cc @ChunyiLyu as I know she had opinions on this whole topic.

Reply:
If the server is under heavy load after the restart, with many messages in queues, it may be better to wait until a quiescent period to rebalance the nodes. I support the idea of feature flagging this, though it may be better to make the default behaviour to rebalance the queues, and require an explicit config to disable the rebalance.

Reply:
Thanks for the great comments 🙂

I also vote for defaulting to run the post-upgrade steps allowing the user to skip the checks if they observe any issues.
Otherwise, I believe few users would enable that flag because for them it's hard to judge when to (not) run post-upgrade steps.